### ISSUE TITLE
Bump minimum Python dependency to 3.7 (or newer)

### ISSUE THREAD

**Issue by huddlej** [nextstrain/cli#136]
## Context

We currently require Python 3.6 or higher, but 3.6 will reach end of life in three weeks.

## Description

We should require at least version 3.7 now.

## Possible solution

 - [ ] Update setup.py to set minimum Python version to 3.7
 - [ ] Update GitHub CI workflow to drop 3.6
 - [ ] Update change log noting breaking change
 - [ ] Release a new major version of CLI to reflect breaking change
 - [ ] [Create a PR](https://github.com/nextstrain/bioconda-recipes/) to update [the minimum Python version in the CLI's bioconda recipe](https://github.com/bioconda/bioconda-recipes/blob/fca9554c95763983bdc9fb06770f956b90967057/recipes/nextstrain-cli/meta.yaml#L20-L24)

Making the change to the bioconda repository first could potentially cause issues for users, although releasing a new CLI version will trigger an autobump of the bioconda package that we will have to supersede with our own PR.

**Comment by tsibley** [comment]
@huddlej and I had [an extensive conversation](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1638397441299000) about Python version bumps in Slack.

TL;DR is that I'd like to keep the CLI at supporting 3.6 for now.

The CLI is the interface to our various runtimes (h/t @victorlin for that clearer term), each of which provide their own isolation levels and Python versions. I'd rather the CLI stayed as minimally restrictive about deps as possible, including the minimum Python version. The idea is that this makes it easier to bootstrap a CLI install, which then gets you access to the runtimes.

**Comment by victorlin** [comment]
That makes sense. From my perspective, the packages we have for consumption (augur, auspice, cli) should be flexible to work with officially maintained versions at the least, while the downstream consumers of these packages (nextstrain.org, docker image) should have deterministic versions and upgraded as appropriate. This is my intent with nextstrain/auspice#1436 and nextstrain/nextstrain.org#438.

Also, I think updating the CI workflows to test on latest supported versions (py3.10 now) is equally beneficial as dropping older versions.

**Comment by tsibley** [comment]
> From my perspective, the packages we have for consumption (augur, auspice, cli) should be flexible to work with officially maintained versions at the least, while the downstream consumers of these packages (nextstrain.org, docker image) should have deterministic versions and upgraded as appropriate.

Definitely agree!

> Also, I think updating the CI workflows to test on latest supported versions (py3.10 now) is equally beneficial as dropping older versions.

Yep, agreed we should track the latest platform versions with our CI.

**Comment by victorlin** [comment]
Some things have happened since this issue was closed:

1. Python 3.7 reached EOL [today](https://devguide.python.org/versions/).

2. A new standalone installer is the preferred method of installation, which (I assume) is kept up to date with an active version of Python.

3. This codebase continues to collect workarounds specific to Python 3.6. Here are the ones I can find:

    - https://github.com/nextstrain/cli/blob/6e36f6d7bdafcb984f62e2a9cc2a9cf9adf0be57/.github/workflows/ci.yaml#L44-L47
    - https://github.com/nextstrain/cli/blob/6e36f6d7bdafcb984f62e2a9cc2a9cf9adf0be57/nextstrain/cli/__main__.py#L29-L31
    - https://github.com/nextstrain/cli/blob/6e36f6d7bdafcb984f62e2a9cc2a9cf9adf0be57/nextstrain/cli/aws/cognito/__init__.py#L7
    - https://github.com/nextstrain/cli/blob/6e36f6d7bdafcb984f62e2a9cc2a9cf9adf0be57/nextstrain/cli/command/build.py#L218-L220
    - https://github.com/nextstrain/cli/blob/6e36f6d7bdafcb984f62e2a9cc2a9cf9adf0be57/nextstrain/cli/util.py#L273-L275
    - https://github.com/nextstrain/cli/blob/6e36f6d7bdafcb984f62e2a9cc2a9cf9adf0be57/setup.py#L87

I'm sure we can keep adding these workarounds, but at what point do we make the call to bump the minimum requirement in favor of cleaner code?

Having more user insight would be ideal. If there are still many users using CLI with Python 3.6, that sounds like a reason to keep active support for it. However, if the number of those users is few (as I'd hope), I think it would be a good idea to drop active support. This doesn't mean that those users can't use Nextstrain CLI anymore, only that they won't be able to use newer versions until they decide to switch to a newer Python version or use the standalone installer.

**Comment by tsibley** [comment]
@victorlin That's a perfect summary of what I've been thinking in the past several months (but not done anything about…).

Requiring 3.8 (or maybe even newer?) would be a welcome change for the codebase IMO, made in small part by the simple passage of time but mostly by the existence of the standalone installation method, which completely eliminates the dependence on an external Python. My biggest reason to cling to supporting older Pythons was to ensure that no matter your compute environment, you could run a recent copy of Nextstrain CLI and thus have access to the managed runtimes. The standalone installs solve that in an even better and more sustainable way than maintaining 3.6 (or 3.7) compat does.

https://nedbatchelder.com/text/which-py.html is a great list of nice features we'll gain for each Python version by setting a new minimum.

**Comment by victorlin** [comment]
I'll put this on my to-do list for sometime within the next month.
